### DISCUSSION TITLE
Difference between Call and Send

### DISCUSSION THREAD

**Discussion by ghost** [ergo-services/ergo#186] [General]
Hi,
just tried to dive into this framework (never had experience with the Erlang/OTP) and faced a strange issue.

I tried to create a chain of synchronous calls based on the generated code (application-supervisor-3 actors, just calling one another: A->B->C with the returning of answer C->B->A). 

But every call is rejected with "not allowed". As I understand, this is due to the [process status](https://github.com/ergo-services/ergo/blob/master/node/process.go#L936). At the same time, the process [should wake up](https://github.com/ergo-services/ergo/blob/master/node/process.go#L1689) if there is something in the mailbox.

I may be missing something, but how can something be added to the mailbox if the very operation of adding to the mailbox checks a status that returns `false` due to the sleeping process because there are no messages in the mailbox?

`Send`, meanwhile, [uses `IsAlive`](https://github.com/ergo-services/ergo/blob/master/node/process.go#L616), which allows the `Sleep` status, and therefore it works stably (but it doesn’t return a response, so it’s not quite suitable for my task).

So, the questions here are next:

Would you kindly to tell me, please, is there any description of the difference between `Send` and `Call`?
And if it is expected behaviour of `Call`, then is there something that should be done to wake up the actor?

Thank you for attention to this matter.

**Comment by halturin** [comment]
Could you please share your code? Its difficult to say whitout seeing it.

Send method allows you to send a message asynchrounously.
Call - to make a synchronous request.

see documentation for the details https://docs.ergo.services/basics/process#process-states

**Reply by ghost** [reply]
> Could you please share your code? 

Sure, there is an [example](https://github.com/4xlan/ergo-test).

>  Call - to make a synchronous request.

So, did I understood correctly then:
1. To make a Call, I need to be sure that goroutine is in `Running` state?
2. Then, if goroutine goes to `Sleep`,  I'm not able to send a synchronous request anymore (`isStateRW`, [which called on the start](https://github.com/ergo-services/ergo/blob/master/node/process.go#L936) of `Call` func [doesn't include `Sleep`](https://github.com/ergo-services/ergo/blob/master/node/process.go#L1706) as the accepted state).

**Reply by halturin** [reply]
You have nothing to do with it

**Reply by halturin** [reply]
seems you didnt get the idea of actors. you shouldnt use goroutines for them. 

here the problem line

https://github.com/4xlan/ergo-test/blob/master/apps/testapp/act1.go#L22

you literally trying to reimplement what actors already have :)

all the methods have to be used whithin the actor's goroutine. if you need to do something periodically, just send a message to itself using SendAfter method.

**Reply by halturin** [reply]
one more thing... https://github.com/4xlan/ergo-test/blob/master/types.go#L18

you dont need to register types for the local messaging. but you willing to send them over the network - yes, all that types have to be registered on both sides - sender and receiver nodes.

if you declared message A on node1, you need to register it on node1. Remote nodes should import type A and register it on their side.

**Reply by ghost** [reply]
Ach, yep, probably it is time to RTFM :)
Shortly, in case if actor is expected to be a part of some constant process (e.g. k8s events listener/db worker/etc.), then all this logic of events listening/work with db shouldn't be implemented in the actor, but actor can be included to this operator's structure and be used to send calls/messages?

**Reply by halturin** [reply]
>Ach, yep, probably it is time to RTFM :)

good idea, btw

**Reply by halturin** [reply]
>Shortly, in case if actor is expected to be a part of some constant process (e.g. k8s events listener/db worker/etc.), then all this logic of events listening/work with db shouldn't be implemented in the actor, but actor can be included to this operator's structure and be used to send calls/messages?

didnt get. could you pls share an example?

**Reply by ghost** [reply]
> didnt get. could you pls share an example?

In case if I have a package, e.g. DB operator, which acts as a separate goroutine, but I'd like to make it react on a messages and provide answer, then I should include the actor into this DB operator's structure (with parent link from actor to DB), and init DB separately from actor?

```go
type DBOp struct {
	actr DBActor
	cfg        *dbConfig
	dbInstance *gorm.DB
}

type DBActor struct {
	act.Actor
        parent *DBOp
}
```

**Reply by halturin** [reply]
to work with something synchronous outside of the actor you need a meta-process https://docs.ergo.services/basics/meta-process (see meta.TCP or meta.UDP as an example).

but to work with DB its enough to have an actor which receives request/messages to make DML requests to your DB.

still not sure if i follow your idea :)

**Reply by ghost** [reply]
> still not sure if i follow your idea :)

Oh... I wish I could follow it by myself
Let me make it more complex then, but I'm afraid that it will go offtopic.

----

Let's assume that we have the some k8s operator, which listens to events in k8s namespace. As it is done with the SDK, we have to operate with the `Reconcile` func, which is called for each event, which fits the conditions.

Inside of this `Reconcile`, it is supposed to perform some operation with the content of the k8s resource, including the storage in the DB. For this purpose, we have two additional packages - content and cache manager. 

Communication chain looks like: k8s op. -> content manager -> cache (db) manager.

In some cases cache manager can return data to content manager (to prepare an update for k8s resource), and content manager can pass it to k8s op. to update the resource in the namespace.

I'm trying now to figure out - is it possible to include the ergo in this chain as a message system. Currently, I have a feeling, that probably I should try to init node/actors in main (without autogenerated structure) and somehow pass the pointers to the structs of exact actors to the structures of k8s op./cache/content mgr.

Probably is there any best practices for this framework, regarding similar cases? Or it is everything about the actor model and the main recommendation is go deeper to model description?

**Reply by halturin** [reply]
I'm not familiar with the k8s internals and its API, but...

I think you can use `gen.Node.Send` method inside of `Reconcile` function to send messages to any process (even to the remote ones). The thing is you can use `gen.Node` interface wherever you want in your code while `gen.Process` interface can be used inside of your Actor callbacks only. 

PS: is `Reconcile` the callback that invoked by SDK internals?

**Reply by halturin** [reply]
you can get `gen.Node` interface on starting your node from `ergo.StartNode(...)` or using `gen.Process.Node()` in your Actor callback (`HandleMessage`, for instance)

**Reply by ghost** [reply]
Understood, will try, thank you!


> PS: is Reconcile the callback that invoked by SDK internals?

Not sure that it is a callback, but this function called for each change in k8s cluster, [there is a description](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/reconcile/reconcile.go#L56).

**Reply by ghost** [reply]
To, probably, finish this topic with some logical ending.

Just tried abovementioned option [here](https://github.com/4xlan/ergo-test/tree/linkexchange).
All these examples above can be shortened to two conditions:
1. Actor should be possible to reach the functions of the "another packet"
2. "Another packet" should be able to call something from the actor.

To reach this goal I had to do some strange message exchange from Svc1 to Act1 and back, to have necessary links on both structs. This exchange works well, I think, because of `Send`'s usage.

Solution itself looks really lousy, and `ProxyCall` was the last chance to check whether this approach is working or not (it was really naive hope that if I will use `Call` from the actor directly, it will solve everything), but, at least, I've checked the theory.

Unfortunately, no luck and I still have this "not allowed" when trying to call the `svc2`. Mayhaps I'm still missing something, will get back to this investigation later.
